src/virtualenv/interpreters/discovery/py_info.py

```diff
@@ -0,0 +1,256 @@
+        # noinspection PyProtectedMember
+        data["version_info"] = data["version_info"]._asdict()  # namedtuple to dictionary
+        return json.dumps(data, indent=2)
+
+    @classmethod
+    def from_json(cls, payload):
+        data = json.loads(payload)
+        data["version_info"] = VersionInfo(**data["version_info"])  # restore this to a named tuple structure
+        info = copy.deepcopy(CURRENT)
+        info.__dict__ = data
+        return info
+
+    @property
+    def system_prefix(self):
+        return self.real_prefix or self.base_prefix or self.prefix
+
+    @property
+    def system_exec_prefix(self):
+        return self.real_prefix or self.base_exec_prefix or self.exec_prefix
+
+    @property
+    def system_executable(self):
+        env_prefix = self.real_prefix or self.base_prefix
+        if env_prefix:  # if this is a virtual environment
+            if self.real_prefix is None and self.base_executable is not None:  # use the saved host if present
+                return self.base_executable
+            # otherwise fallback to discovery mechanism
+            return self.find_exe_based_of(inside_folder=env_prefix)
+        else:
+            # need original executable here, as if we need to copy we want to copy the interpreter itself, not the
+            # setup script things may be wrapped up in
+            return self.original_executable
+
+    def find_exe_based_of(self, inside_folder):
+        # we don't know explicitly here, do some guess work - our executable name should tell
+        possible_names = self._find_possible_exe_names()
+        possible_folders = self._find_possible_folders(inside_folder)
+        for folder in possible_folders:
+            for name in possible_names:
+                candidate = os.path.join(folder, name)
+                if os.path.exists(candidate):
+                    info = PythonInfo.from_exe(candidate)
+                    keys = {"implementation", "architecture", "version_info"}
+                    if all(getattr(info, k) == getattr(self, k) for k in keys):
+                        return candidate
+        what = "|".join(possible_names)  # pragma: no cover
+        raise RuntimeError("failed to detect {} in {}".format(what, "|".join(possible_folders)))  # pragma: no cover
+
+    def _find_possible_folders(self, inside_folder):
+        candidate_folder = OrderedDict()
+        base = os.path.dirname(self.executable)
+        # following path pattern of the current
+        if base.startswith(self.prefix):
+            relative = base[len(self.prefix) :]
+            candidate_folder["{}{}".format(inside_folder, relative)] = None
+
+        # or at root level
+        candidate_folder[inside_folder] = None
+        return list(candidate_folder.keys())
+
+    def _find_possible_exe_names(self):
+        name_candidate = OrderedDict()
+        for name in [self.implementation, "python"]:
+            for at in range(3, -1, -1):
```

Comment:
```suggestion
            for at in (3, 2, 1, 0):
```

slightly more readable -- but now that I've expanded this i'm confused what it's trying to do? do we expect to see `python0` executables? is this going to break if/when python4?

Reply:
the at is used to count the number of version parts to use; e.g. ``python3.8.1, python3.8, python3, python``